### ISSUE TITLE
Compile errors on Mac with M1 chip - This macro cannot be used on the current target.

### ISSUE THREAD

**Issue by apps4uco** [liebharc/basic_dsp#49]
I get compile errors on a Mac with Apple Silicon

 rustc --version
rustc 1.65.0-nightly (86c6ebee8 2022-08-16)

uname -a                        
Darwin andy-mac-m1.local 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:20:05 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T8101 arm64

~~~
Compiling basic_dsp_vector v0.9.4
error: This macro cannot be used on the current target.
                                   You can prevent it from being used in other architectures by
                                   guarding it behind a cfg(any(target_arch = "x86", target_arch = "x86_64")).
  --> /Users/___/.cargo/registry/src/-ea8217003c7340b4/basic_dsp_vector-0.9.4/src/vector_types/support_std.rs:21:19
   |
21 |     let reg_len = sel_reg!(get_reg_len::<f32>());
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

..... 87 similar errors cut

error: could not compile `basic_dsp_vector` due to 88 previous errors
~~~

Btw I would be happy to test any fix

**Comment by liebharc** [comment]
Hi @apps4uco ,

I don't have Apple hardware myself, and the M1 would be beyond my budget. Also, I'm not really in touch with the Rust ecosystem since two years or so.  I believe we can get the compile error fixed, but it will likely require a lot of trial and error. So please be prepared for that if you're offering to test this :). 

That said, I pushed a change to master which COULD help. Really not sure if it does. Rust does an excellent job again in providing good error messages, but without any means to test it myself, it's still kind of a blind guess.

**Comment by apps4uco** [comment]
I cloned the project and still get the same 88 errors.

But I got an idea from your fix, however, I then discovered that #[cfg] in a macro isnt so easy to implement.
So I ended commenting out all the Simd tests just to get something working.

```
/// Selects a SIMD register type and passes it as 2nd argument to a function.
/// The macro tries to mimic the Rust syntax of a method call.
macro_rules! sel_reg(
    ($self_:ident.$method: ident::<$type: ident>($($args: expr),*)) => {
                // TODO enable AVX512 detection again as soon as stdsimd is stable
        //if is_x86_feature_detected!("avx512vl") && cfg!(feature="use_avx512") {
//            $self_.$method(RegType::<<$type as ToSimd>::RegAvx512>::new(), $($args),*)
//#[cfg(all(feature = "use_avx2",target_feature="avx2")]
//            $self_.$method(RegType::<<$type as ToSimd>::RegAvx>::new(), $($args),*)
//#[cfg(all(feature = "use_sse2",target_feature="sse2")]
//            $self_.$method(RegType::<<$type as ToSimd>::RegSse>::new(), $($args),*)
            $self_.$method(RegType::<<$type as ToSimd>::RegFallback>::new(), $($args),*)
    };
    ($method: ident::<$type: ident>($($args: expr),*)) => {
        //if is_x86_feature_detected!("avx512vl") && cfg!(feature="use_avx512") {
//            $method(RegType::<<$type as ToSimd>::RegAvx512>::new(), $($args),*)
//#[cfg(all(feature = "use_avx2",target_feature="avx2")]
//            $method(RegType::<<$type as ToSimd>::RegAvx>::new(), $($args),*)
//#[cfg(all(feature = "use_sse2",target_feature="sse2")]
//            $method(RegType::<<$type as ToSimd>::RegSse>::new(), $($args),*)
            $method(RegType::<<$type as ToSimd>::RegFallback>::new(), $($args),*)
    };
);

```

Doing that all the tests pass on the Mac M1.

The following code is my suggestion for a fix but it has an error and Im not sure how to fix it.

```
/// Selects a SIMD register type and passes it as 2nd argument to a function.
/// The macro tries to mimic the Rust syntax of a method call.
macro_rules! sel_reg(
    ($self_:ident.$method: ident::<$type: ident>($($args: expr),*)) => { 
                // TODO enable AVX512 detection again as soon as stdsimd is stable
        //if is_x86_feature_detected!("avx512vl") && cfg!(feature="use_avx512") {
//            $self_.$method(RegType::<<$type as ToSimd>::RegAvx512>::new(), $($args),*)
#[cfg(target_feature = "avx2")]
            $self_.$method(RegType::<<$type as ToSimd>::RegAvx>::new(), $($args),*)
#[cfg(all(target_feature = "sse2", not(target_feature = "avx2")))]
            $self_.$method(RegType::<<$type as ToSimd>::RegSse>::new(), $($args),*)
#[cfg(not(any(target_feature = "sse2", target_feature = "avx2")))]
            $self_.$method(RegType::<<$type as ToSimd>::RegFallback>::new(), $($args),*)
    };
    ($method: ident::<$type: ident>($($args: expr),*)) => {
        //if is_x86_feature_detected!("avx512vl") && cfg!(feature="use_avx512") {
//            $method(RegType::<<$type as ToSimd>::RegAvx512>::new(), $($args),*)
#[cfg(target_feature = "avx2")]
            $method(RegType::<<$type as ToSimd>::RegAvx>::new(), $($args),*)
#[cfg(not(any(target_feature = "sse2", target_feature = "avx2")))]
            $method(RegType::<<$type as ToSimd>::RegSse>::new(), $($args),*)
#[cfg(not(any(target_feature = "sse2", target_feature = "avx2")))]
            $method(RegType::<<$type as ToSimd>::RegFallback>::new(), $($args),*)
    };
);

```
I get a suggestion to put ; at the end of the lines but when I do that I get

warning: trailing semicolon in macro used in expression position

Hope it helps

**Comment by liebharc** [comment]
Thanks, yes it helps. I think it confirms that we're looking at the right spot in the code and that the general idea of fixing this is right.

An explanation of what I tried: I was assuming that the `cfg` macro is enough to make the compiler ignore the remaining conditions and thus never try to compile the `is_x86_feature_detected` macro (from what I read the M1 has no SSE or AVX support, so those checks should return false). From the docs, it's clear that I was wrong there:

```
cfg!, unlike #[cfg], does not remove any code and only evaluates to true or false. 
```

https://doc.rust-lang.org/std/macro.cfg.html

Essentially, we want to achieve for now what you did by removing the if checks: The neon implementation should just use the fallback reg types, which means that we only get the vector optimization which Rust adds automatically. From the benchmarks I did some years ago, the Rust compiler did for a lot of operations a good job on its own. However, different to your changes, I'd like to still support the vector optimizations for x86 CPUs as that's already there.

In the long run, it might be desirable to support the neon vector types, but that would need someone with a M1 CPU to be willing to commit quite some time.
